modules/3d/src/rendering.cpp

```diff
@@ -0,0 +1,202 @@
+                    if (z_interpolated < depth_buf[index] && z_interpolated >= 0.0 && z_interpolated <= 1.0)
+                    {
+                        if (isConstant)
+                            color_buf[index] = tri.getTriangleColor();
+                            //color_buf[index] = Vec3f(255, 0, 0);
+                        else {
+                            float r1 = alpha * z_interpolated / tri.vertices[0][2] * tri.color[0][0] + beta * z_interpolated / tri.vertices[1][2] * tri.color[1][0]
+                                + gamma * z_interpolated / tri.vertices[2][2] * tri.color[2][0];
+                            float g1 = alpha * z_interpolated / tri.vertices[0][2] * tri.color[0][1] + beta * z_interpolated / tri.vertices[1][2] * tri.color[1][1]
+                                + gamma * z_interpolated / tri.vertices[2][2] * tri.color[2][1];
+                            float b1 = alpha * z_interpolated / tri.vertices[0][2] * tri.color[0][2] + beta * z_interpolated / tri.vertices[1][2] * tri.color[1][2]
+                                + gamma * z_interpolated / tri.vertices[2][2] * tri.color[2][2];
+
+                            color_buf[index] = Vec3f(r1, g1, b1);
+                        }
+                        depth_buf[index] = z_interpolated;
+                    }
+                }
+            }
+    }
+
+    void triangleRasterize(const std::vector<Vec3f>& vertices, const std::vector<Vec3i>& indices,
+                           const std::vector<Vec3f>& colors, const Vec3f& position, const Vec3f& lookat, const Vec3f& upVector,
+                           float fovy, float zNear, float zFar, int width, int height, bool isConstant,
+                           std::vector<float>& depth_buf, std::vector<Vec3f>& color_buf)
+    {
+        Matx44f lookAtMatrix = lookAtMatrixCal(position, lookat, upVector);
+        Matx44f perspectMatrix = perspectMatrixCal((float)width / (float)height, fovy, zNear, zFar);
+        Matx44f mvpMatrix = perspectMatrix * lookAtMatrix;
+
+        depth_buf.resize(width * height);
+        color_buf.resize(width * height);
+
+        std::fill(depth_buf.begin(), depth_buf.end(), std::numeric_limits<float>::infinity());
+        std::fill(color_buf.begin(), color_buf.end(), Vec3f(0.0, 0.0, 0.0));
+
+        for (int i = 0; i < indices.size(); i++)
+        {
+            Triangle tri;
+            Vec4f ver[3] = {
+                mvpMatrix * Vec4f(vertices[indices[i][0]][0], vertices[indices[i][0]][1], vertices[indices[i][0]][2], 1.0),
+                mvpMatrix * Vec4f(vertices[indices[i][1]][0], vertices[indices[i][1]][1], vertices[indices[i][1]][2], 1.0),
+                mvpMatrix * Vec4f(vertices[indices[i][2]][0], vertices[indices[i][2]][1], vertices[indices[i][2]][2], 1.0)
+            };
+
+            for (auto& vertex: ver)
+            {
+                vertex = Vec4f(vertex[0] / vertex[3], vertex[1] / vertex[3], vertex[2] / vertex[3], 1.0);
```

Comment:
This exact line can cause perspective correction issues (the difference between RGB image you get and OpenGL's ground truth can be explained by this: you need Z values to interpolate, while you're passing 1's instead)